src/imageio/format/jxl.c

```diff
@@ -0,0 +1,749 @@
+        // TODO: safe to simply put pipe->icc_intent here?
+        color_encoding.rendering_intent = JXL_RENDERING_INTENT_PERCEPTUAL;
+
+        JxlEncoderSetColorEncoding(encoder, &color_encoding);
+      }
+    }
+  }
+
+  if(write_icc)
+  {
+    fprintf(stderr, "JXL: Could not generate color encoding data, falling back to ICC binary\n");
+
+    cmsUInt32Number size = 0;
+    if(cmsSaveProfileToMem(out_profile, NULL, &size))
+    {
+      unsigned char *buf = malloc(size);
+      if(!buf || !cmsSaveProfileToMem(out_profile, buf, &size)) goto end;
+
+      JXL_ASSERT(JxlEncoderSetICCProfile(encoder, buf, size));
+
+      free(buf);
+    }
+    else
+    {
+      fprintf(stderr, "JXL: Error writing ICC data\n");
+      goto end;
+    }
+  }
+
+  JxlBasicInfo basic_info;
+  JxlEncoderInitBasicInfo(&basic_info);
+  basic_info.have_container = JXL_FALSE;
+  basic_info.xsize = width;
+  basic_info.ysize = height;
+  basic_info.bits_per_sample = 32;
+  basic_info.exponent_bits_per_sample = 8;
+  basic_info.intensity_target = 255.f;
+  basic_info.min_nits = 0.f;
+  basic_info.relative_to_max_display = JXL_FALSE;
+  basic_info.linear_below = 0.f;
+  basic_info.uses_original_profile = JXL_TRUE;
+  basic_info.have_preview = JXL_FALSE;
+  basic_info.have_animation = JXL_FALSE;
+  basic_info.orientation = JXL_ORIENT_IDENTITY;
+  basic_info.num_color_channels = 3;
+  basic_info.num_extra_channels = 0;
+  basic_info.alpha_bits = 0;
+  basic_info.alpha_exponent_bits = 0;
+  basic_info.alpha_premultiplied = JXL_FALSE;
+  JXL_ASSERT(JxlEncoderSetBasicInfo(encoder, &basic_info));
+
+  // We assume that the user wants the JXL image in a bmff container.
+  // JXL images can be stored without any container so they are smaller, but this
+  // removes the possibility of storing extra data like exif.
+  // It is not likely darktable users exporting images care so much about this
+  // small size improvement, and more likely they want to store exif data.
+  // HOWEVER - We do not want libjxl to write the image in a container, we will
+  // do that ourselves. This gives us the control to be able to add our own
+  // sections (e.g. exif data).
+  JXL_ASSERT(JxlEncoderUseContainer(encoder, JXL_FALSE));
+
+  // automatically freed when we destroy the encoder
+  JxlEncoderOptions *options = JxlEncoderOptionsCreate(encoder, NULL);
+  JXL_ASSERT(JxlEncoderOptionsSetLossless(options, params->lossless ? JXL_TRUE : JXL_FALSE));
+
+  int decode_effort = params->decoding_effort;
+  // We store decoding effort but jxl wants encoding speed, we must reverse it
+  decode_effort = (decode_effort - 4) * -1;
+  JXL_ASSERT(JxlEncoderOptionsSetDecodingSpeed(options, decode_effort));
+
+  const int encode_effort = params->encoding_effort;
+  JXL_ASSERT(JxlEncoderOptionsSetEffort(options, encode_effort));
+
+  int quality = params->quality;
+  // We store quality but jxl wants distance, so we reverse the scale
+  quality = (quality - 15) * -1;
```

Comment:
Same as above, let's stick to the API and cjxl options, as that is what people will refer to.

Reply:
Good idea.

By the way, Thank You so much for your helpful reviews! Getting through them slowly! :sweat_smile:

Reply:
Np, hope I'm not 'nagging' too much, keep up the great work ;)

Reply:
0 distance and lossless mode is not the same thing.